docs/keyed_services.md

```diff
@@ -0,0 +1,61 @@
+### Keyed Services
+
+Most features should be scoped to a Profile/BrowserContext and be BrowserContextKeyedServices per https://www.chromium.org/developers/design-documents/profile-architecture/. What this means is that all associated preferences and other locally stored state will be per-profile. A service may sometimes be shared between the regular and OTR profile (check with sec-team) and it may also be ununavailable for certain profile types. `ProfileKeyedServiceFactory/ProfileKeyedServiceFactoryIOS` with `ProfileSelections::Builder` is the preferred method to use for determining which (if any) profile should be used for the service.
+
+```
+MyServiceFactory::MyServiceFactory()
+    : ProfileKeyedServiceFactory(
+          "MyService",
+          ProfileSelections::Builder()
+              .WithRegular(ProfileSelection::kOriginalOnly)
+              .Build()) { ... }
+```
+
+If you have more complicated logic then use `BrowserContextKeyedServiceFactory::GetBrowserContextToUse` and return nullptr if the service is not available for the given `BrowserContext`.
+
+```
+content::BrowserContext* MyServiceFactory::GetBrowserContextToUse(
+    content::BrowserContext* context) const {
+  Profile* profile = Profile::FromBrowserContext(context);
+  return profile->IsRegularProfile() &&
+         base::FeatureList::IsEnabled(features::kMyFeature);
+}
+```
+
+Do not explicitly return null from other methods like `BuildServiceInstanceForBrowserContext`, `GetForProfile`, etc... If you return nullptr from `GetBrowserContextToUse` or configure `ProfileSelections`, the service will already be null. Avoid using helper methods like `IsAllowedForProfile` to see if a service is available for a particular profile as these are often used incorrectly/inconsistently. The preferred method is just a null check on the service itself. Remember to prefer dependency injection https://chromium.googlesource.com/chromium/src/+/main/docs/chrome_browser_design_principles.md#structure_modularity and pass in the services you require instead of calling the factory methods internally. This reduces dependencies on the factories and makes unit testing/mocking simpler.
```

Comment:
re `IsAllowedForProfile` - we do need some way to make sure checks are consistent between desktop and ios factories, but it's tricky because ios subclasses `web::BrowserState` instead of `BrowserContext`. Maybe we should have profile type constants like kRegular, kIncognito, kTor, etc... that we pass to a static method on the keyed service along with anything else we might need like `PrefService`?

Reply:
ios has `ProfileIOSType` and desktop has some for metrics, but we also need brave-specific profile types